examples/lighting-app/lighting-common/DataModelHandler.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *   This file implements the handler for data model messages.
+ */
+
+#include <lib/support/logging/CHIPLogging.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+#include "DataModelHandler.h"
+#include "LightingManager.h"
+
+#include "af-types.h"
+#include "attribute-storage.h"
+#include "gen/attribute-id.h"
+#include "gen/cluster-id.h"
+#include "gen/znet-bookkeeping.h"
+#include "util.h"
+#include <app/chip-zcl-zpro-codec.h>
+
+using namespace ::chip;
+
+extern "C" {
+void emberAfPostAttributeChangeCallback(uint8_t endpoint, EmberAfClusterId clusterId, EmberAfAttributeId attributeId, uint8_t mask,
```

Comment:
if one removes this function, can this file be moved to a library?

Reply:
Yes, I believe it can....  and then it can be shared with the esp32 app as well, actually.  I'm OK with that happening as a followup.  @erjiaqing if you're willing to do that, let me know; if not, I can do it.

Reply:
I have no idea about removing this. For now, different apps have different callback handler, and one app might handle many clusters here. so I still treat it as app specified.

Or, do you mean by using macros to define all callbacks like (or some other methods that are similar to) this?

```cpp
switch (clusterId) {
#ifdef ENABLE_ON_OFF_CLUSTER
case ZCL_ON_OFF_CLUSTER_ID:
 {...  OnOffChangeCallback(endpoint, value);}
#endif
#ifdef ENABLE_LEVEL_CLUSTER
case ZCL_LEVEL_CLUSTER_ID:
 {...  LevelChangeCallback(endpoint, value);}
#endif
```

Reply:
In the SiLabs model, this function is supplied by the implementation, is part of the interface to the "application", that knows which attributes to watch to twiddle GPIOs.

We can move DataModelHandler.cpp into chip and have it call this named function, which would be supplied by each example app.

Reply:
I checked the code, and think this is necessary and possible, but before we can move DataModelHandler into CHIP core, we need to move some code from `gen` to `app`. This can be a follow up.